### DISCUSSION TITLE
Abstracting common fields into one abstract entity

### DISCUSSION THREAD

**Discussion by Juknum** [mikro-orm/mikro-orm#4629] [Q&A] [answered]
Hello,

I'm trying to make a file management "system" for my project with view groups to determine if a user has access or not to a file;

I'm currently doing this:
```ts
@Entity({ abstract: true })
export abstract class File {
	// ... Others fields

	@Property()
	@ApiProperty()
	filename: string;

	@ManyToOne(() => FileVisibilityGroup, { nullable: true })
	@ApiProperty({ type: Number, minimum: 1 })
	visibility?: FileVisibilityGroup;
}
```

```ts
@Entity({ tableName: 'files_visibility_groups' })
export class FileVisibilityGroup extends BaseEntity implements FileVisibilityGroupEntity<User> {
	@Property()
	@ApiProperty()
	name: Uppercase<string>;

	@Property()
	@ApiProperty()
	description: string;

	//* Note: Use the 'string' version of the entity name to avoid circular dependency issues.
	@ManyToMany(() => 'User', (user: User) => user.files_visibility_groups, { owner: true, nullable: true })
	users = new Collection<User>(this);

	@OneToMany(() => File, (file) => file.visibility, { nullable: true })
	files = new Collection<File>(this);
}
```

And I'm extending the `File` abstract class into each entity I need : 
```ts
@Entity({ tableName: 'promotions_pictures' })
export class PromotionPicture extends File {
	@ApiProperty({ type: Number, minimum: 1 })
	@OneToOne(() => Promotion, (promotion) => promotion.picture, { owner: true, unique: true })
	promotion: Promotion;
}
```
```ts
@Entity({ tableName: 'users_pictures' })
export class UserPicture extends File implements UserPictureEntity<User> {
	@ApiProperty({ type: Number, minimum: 1 })
	@OneToOne(() => User, (user) => user.picture, { owner: true, unique: true })
	user: User;
}
```

And my user entity looks like this: 

```ts
@Entity({ tableName: 'users' })
export class User {
	// ... Others fields

	@OneToOne(() => UserPicture, (picture) => picture.user, { cascade: [Cascade.ALL], nullable: true })
	@ApiProperty()
	picture?: UserPicture;

	@ManyToMany(() => FileVisibilityGroup, (group) => group.users, { nullable: true })
	files_visibility_groups? = new Collection<FileVisibilityGroup>(this);
```

But then, when I'm creating a `FileVisibilityGroup` entity using `em.create(FileVisibilityGroups, { name: "TEST", description: "" })` I get the following error: `Metadata for entity File not found`. I guess I did something wrong but I don't know if what I'm doing should be done like that aha ^^"

> TL:DR; The main idea is to get an abstracted entity for each files (could be an image or a PDF or whatever) which would have a "visibility" field to allow certain people to access the file, but I can't get it to work properly :'(

> PS: I'm using NestJS & PostgreSQL

**Comment by B4nan** [comment]
You can't target an abstract entity in any relation, they don't represent any database table, so they cannot be used to form a foreign key constraint.

**Reply by Juknum** [reply]
So it's not possible to reference different entities in a collection ? Should I make a FileVisibilityGroup for each child of File ?

**Reply by Juknum** [reply, marked as answer]
Finally found what I was looking for : https://mikro-orm.io/docs/inheritance-mapping
